core/src/main/java/cucumber/runtime/ThreadLocalRunnerSupplier.java

```diff
@@ -29,16 +34,11 @@ public ThreadLocalRunnerSupplier(
 
     @Override
     public Runner get() {
-        Runner runner = runners.get();
-        if (runner == null) {
-            runner = createRunner();
-            runners.set(runner);
-        }
-        return runner;
+        return runners.get();
     }
 
     private Runner createRunner() {
-        return new Runner(glueSupplier.get(), eventBus.createBatchedEventBus(), backendSupplier.get(), runtimeOptions);
+        return new Runner(glueSupplier.get(), eventBus, backendSupplier.get(), runtimeOptions);
```

Comment:
> There is one new test cucumber.runtime.junit.CucumberTest#cucumber_can_run_pickles_in_parallel which is now failing. I don't quite understand the problem at the moment - could look at this another day but think you know this better than I

I reckon this has something to do with it.

Reply:
OK, but I had to get rid of that in the call, as from certain tests I had to pass in a test specific EventBus so that the time on the stopwatch was incremented upon each `TestStepStarted` event

Reply:
Oh and I debugged `BatchedEventBus` and the `super.send()` indeed did nothing as that bus itself had no listeners registered - but it's parent (delegate) did

So i'd be surprised if this was the culprit, but am happy(ish 😉 ) to be wrong

Reply:
>  I had to get rid of that in the call, as from certain tests I had to pass in a test specific EventBus so that the time on the stopwatch was incremented upon each TestStepStarted event

That is why the `TestHelper` was using a different setup.

Reply:
> So i'd be surprised if this was the culprit, but am happy(ish wink ) to be wrong

So what happens:

We use JUnits ParallelComputer to run two features in parallel. For each features we connect the event bus to the JUnits RunNotifier.  

https://github.com/cucumber/cucumber-jvm/blob/master/junit/src/main/java/cucumber/runtime/junit/PickleRunners.java#L147

Because the bus is now shared between threads, all notifiers listen to the same bus and will not only receive events from the execution of their own feature, but also from the other feature. These notifiers send their events up the chain of notifiers, which results in one event being reported multiple times (and/or other concurrent shenanigans).

Reply:
So each runner should have it's own event bus. Like so:

```
 Junit ---listens to---> [TestCase EventBus for Runner 1] ---sends batches of events to--> [Synchronized EventBus]  <---listens too--- HTML formatter 
 Junit ---listens to---> [TestCase EventBus for Runner 2] ---sends batches of events to--> [Synchronized EventBus]  <---listens too--- HTML formatter 
```

Reply:
`So each runner should have it's own event bus. Like so:`

ah yes, sorry you explained that before and I'd forgotten.

Have just fixed the test and pushed

Reply:
> A SynchronizedEventBus wasn't required as a delegate as each thread has it's own list of Formatters to notify of the queued events for the current thread

How does that work? Say run cucumber with 5 runners, each on their own thread, how do we end up with a single json file?

Reply:
Hmm, sounds like there is a problem with the test in that there aren’t an plugins being verified.

It could be that even though the test passes (haven’t changed it) that it isn’t correctly asserting/verifying what should be happening, and thus it is now not producing the desired outcome

Will check this out, but probably not until tomorrow now.